### PR TITLE
Remove TopologicalSortUtility.GetParentCount allocs

### PR DESCRIPTION
**Pull request by benaadams**
Resolves https://github.com/NuGet/Home/issues/5666

For `nuget restore Roslyn.sln -NoCache`

`TopologicalSortUtility` allocates 4,833 MB; taking 21,577ms (21.5 secs)

`TopologicalSortUtility` allocates 16 MB; taking 338ms

For 1024 dependencies its x 90 times faster and only uses 0.7% of the memory
```
                 Method | Deps |        Mean | Scaled |    Gen 0 |  Gen 1 | Gen 2 |    Allocated |
----------------------- |----- |------------:|-------:|---------:|-------:|------:|-------------:|
 TopologicalSortCurrent | 1024 | 8,047.69 ms |   1.00 | 16125.00 | 768.52 | 62.50 | 49,720.89 KB |
 TopologicalSortReduced | 1024 | 2,018.55 ms |   0.25 |    62.50 |      - |     - |    353.02 KB |
  TopologicalSortLookup | 1024 |    88.84 ms |   0.01 |   125.00 |      - |     - |    476.80 KB |
 TopologicalSortLookup2 | 1024 |    89.90 ms |   0.01 |   125.00 |      - |     - |    388.15 KB |
```

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (sharwell) [inline src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: :bulb: This `foreach` was fine before.

File: src/NuGet.Core/NuGet.Packaging.Core/PackageDependencyInfo.cs
Comment (sharwell) [inline src/NuGet.Core/NuGet.Packaging.Core/PackageDependencyInfo.cs]: :bulb: `_dependencies?.ToArray() ?? _empty`

💭 Did `?.` support extension methods from the start? Now I forget.

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (sharwell) [inline src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: 💭 This is pretty dirty / prone to regress. If we use this approach for a short-term resolution we should still follow-up with a proper solution (modifying the `Dependencies` property to return a type with a struct enumerator and/or adding a `PackageDependencyInfo.ContainsDependency` method).

**Review by sharwell — APPROVED** [review]

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: `foreach` on a List is ugly; the enumerator is quite large and keeps a version to watch for changes

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: Agreed; but will leave to maintainers to change api shape :)

File: src/NuGet.Core/NuGet.Packaging.Core/PackageDependencyInfo.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging.Core/PackageDependencyInfo.cs]: Changed

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (sharwell) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: :memo: By follow-up I mean file a separate issue in NuGet/Home. We fixed one thing, but introduced another, so there needs to be an issue filed for the latter.

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (sharwell) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: It won't make a difference in the scheme of things here. The `foreach` is cleaner and less prone to bugs; I'd prefer to use it unless it's observed to have non-negligible overhead in practice (keep in mind I'm testing this on _very_ large projects, so if it doesn't appear as a problem in my testing then likely no one will be able to find it).

**Comment by benaadams** [comment]
@sharwell reworked it so it doesn't allocate via Linq per loop

**Comment by benaadams** [comment]
`Failed   NuGet.Commands.Test.IncludeTypeTests.IncludeType_MultipleDependenciesWithTargets` 😄

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (benaadams) [inline src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: As @sharwell mentions likely need to raise an issue for this for an Api change rather than an `as` cast?

Comment (emgarten) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: I understand that this is casting if it is already an array, or making one. But does casting it here cause foreach to treat it differently than if it is an array behind the IEnumerable interface? How does this work?

Comment (sharwell) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: @emgarten When you use `foreach` on an array, the compiler emits the following, bypassing `IEnumerable` altogether:

```csharp
var tempArray = [original expression];
for (int tempVariable = 0; tempVariable < tempArray.Length; tempVariable++)
{
  [foreach variable] = tempArray[tempVariable];
  [foreach body]
}
```

File: src/NuGet.Core/NuGet.Packaging.Core/PackageDependencyInfo.cs
Comment (alpaix) [inline src/NuGet.Core/NuGet.Packaging.Core/PackageDependencyInfo.cs]: `_empty` => `EmptyDependencies`

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (alpaix) [inline src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: `_comparer` => `DefaultComparer`

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (alpaix) [inline src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: `PackageInfoComparer`

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: It allocates [behind IEnumerable](https://gist.github.com/benaadams/66f736fd0a0fce5699863e4c9e69e851) and is and order of magnitude slower 
```
                  Method |      Mean |     Error |    StdDev | Allocated |
------------------------ |----------:|----------:|----------:|----------:|
                ForArray |  2.694 us | 0.0057 us | 0.0054 us |       0 B |
            ForEachArray |  2.213 us | 0.0081 us | 0.0075 us |       0 B |
           ForArrayIList | 12.068 us | 0.0393 us | 0.0368 us |       0 B |
       ForEachArrayIList | 25.348 us | 0.0851 us | 0.0796 us |      32 B |
 ForEachArrayIEnumerable | 25.308 us | 0.0878 us | 0.0821 us |      32 B |
               LinqArray | 31.287 us | 0.1115 us | 0.1043 us |      32 B |
```
A regular `for` loop with `IList<T>` and a cached length (so it doesn't evaluate the property each loop) is ok (rather than `foreach`) if you want to do an interface e.g.

```csharp
public int ForArrayIList()
{
    var counter = 0;
    var list = _ilist;
    
    var count = list.Count;
    for (var i = 0; i < count; i++)
    {
        counter += list[i];
    }
    return counter;
}
```

**Comment by benaadams** [comment]
Renames + rebased

File: src/NuGet.Core/NuGet.Packaging.Core/PackageDependencyInfo.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging.Core/PackageDependencyInfo.cs]: renamed

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: renamed

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: renamed

**Review by alpaix — APPROVED** [review]
LGTM. Will have to see all tests passed though

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: Allocations aside; interfaces don't inline and have about ~2x the overhead of a regular non-inlined call. Which is fine normally as its quite a small overhead - though not great for tight loops.

And enumeration is a double call `MoveNext` + `Current` so ~x4

Comment (emgarten) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: TIL, thank you @benaadams and @sharwell :smile:

**Review by emgarten — APPROVED** [review]
Thanks so much for the PR, this looks great! 🚀

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (alpaix) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: What's the relation of N and the allocation/time difference? Is it fixed one-time allocation per entire clause or per single run?
Also, the benchmark compares times for pre-allocated arrays and it doesn't cover initialization cost.
For instance, here at the line 20 `ToArray` statement may be more expensive than list vs. array for-loop perf gain.

**Comment by alpaix** [comment]
We should consider adding benchmark/perf test for `TopologicalSortUtility`.

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: > For instance, here at the line 20 ToArray statement may be more expensive than list vs. array for-loop perf gain.

Previously it was doing `.ToList()` and `.ToArray()` is approximately equivalent. 

There is only `1` array allocation for `N` items; 
whereas previously it would be `N + (1 + N) / 2` allocations as it allocates per loop. 

Will put together a benchmark.

**Comment by benaadams** [comment]
[TopologicalSort.cs](https://gist.github.com/benaadams/3e9298fbc4f1f3a48ae80de39e1225b2)

```
                 Method | Deps |             Mean | Scaled |     Gen 0 |   Gen 1 |    Allocated |
----------------------- |----- |-----------------:|-------:|----------:|--------:|-------------:|
 TopologicalSortCurrent |    1 |         992.9 ns |   1.00 |    0.3719 |       - |      1,176 B |
 TopologicalSortReduced |    1 |         748.3 ns |   0.75 |    0.2642 |       - |        832 B |
 TopologicalSortCurrent |    2 |       1,806.9 ns |   1.00 |    0.5684 |       - |      1,792 B |
 TopologicalSortReduced |    2 |       1,271.2 ns |   0.70 |    0.3452 |       - |      1,088 B |
 TopologicalSortCurrent |    4 |       4,023.5 ns |   1.00 |    1.0452 |       - |      3,312 B |
 TopologicalSortReduced |    4 |       2,410.6 ns |   0.60 |    0.4768 |       - |      1,512 B |
 TopologicalSortCurrent |    8 |      12,305.8 ns |   1.00 |    2.5177 |       - |      7,944 B |
 TopologicalSortReduced |    8 |       5,903.3 ns |   0.48 |    0.9384 |       - |      2,976 B |
 TopologicalSortCurrent |   16 |      52,690.3 ns |   1.00 |    6.8970 |       - |     21,760 B |
 TopologicalSortReduced |   16 |      18,723.9 ns |   0.36 |    1.8921 |       - |      6,000 B |
 TopologicalSortCurrent |   32 |     325,924.8 ns |   1.00 |   21.4844 |       - |     67,768 B |
 TopologicalSortReduced |   32 |      93,117.1 ns |   0.29 |    3.6621 |       - |     11,744 B |
 TopologicalSortCurrent |   64 |   2,101,326.4 ns |   1.00 |   70.3125 |       - |    233,456 B |
 TopologicalSortReduced |   64 |     592,494.9 ns |   0.28 |    6.8359 |       - |     23,064 B |
 TopologicalSortCurrent |  512 | 958,608,363.5 ns |   1.00 | 4062.5000 | 62.5000 | 12,874,392 B |
 TopologicalSortReduced |  512 | 250,287,988.2 ns |   0.26 |         - |       - |    181,016 B |
```
Takes a long time to benchmark... O_o

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (agocke) [inline src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: ToList seems better than ToArray here because it shouldn't require an additional copy allocation.

Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: You mean the `Select` transform? That's to add `ParentCount` as a property for caching the parent count so it only needs to be allocated once per function call, rather than a per loop key and map allocation via Linq.Sort (which is still better than a pure functional compare with parent recalc per sort element, without caching the values)

Also since `PackageInfo` is a struct; as a bonus, it changes stops the underlying array (in List or direct) from being covariant and having per op checks, for an extra boost 😉

Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: Measurements are earlier comment above

TopologicalSortCurrent == current code
TopologicalSortReduced == this PR

For a single entry, so single loop iteration allocates less:
```
                 Method | Deps |     Gen 0 |   Gen 1 |    Allocated |
----------------------- |----- |----------:|--------:|-------------:|
 TopologicalSortCurrent |    1 |    0.3719 | 62.5000 |      1,176 B |
 TopologicalSortReduced |    1 |    0.2642 |       - |        832 B |
```
Increasing the dependencies, you can see the allocations are definitely on different growth curves
```
                 Method | Deps |     Gen 0 |   Gen 1 |    Allocated |
----------------------- |----- |----------:|--------:|-------------:|    
 TopologicalSortCurrent |  512 | 4062.5000 | 62.5000 | 12,874,392 B |
 TopologicalSortReduced |  512 |         - |       - |    181,016 B |
```

Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: Looking at in a more traditional profiler, you can see the top level function `SortPackagesByDependencyOrder` allocations are reduced, and the subfunction `GetParentCount` allocations are eliminated:



Looking at execution time; you can see most of the time is spent in the subfunction `GetParentCount` and that is dramatically reduced; while even the top level function's self time is also reduced:

Comment (benaadams) [inline src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: Hmm... we can go much faster... give me a few...

**Comment by benaadams** [comment]
Hmm... we can go much faster... give me a few...

**Comment by benaadams** [comment]
Little more allocations than the reduced version; but faster and better scaling for more dependencies
```
                 Method | Deps |               Mean | Scaled |     Gen 0 |   Gen 1 |    Allocated |
----------------------- |----- |-------------------:|-------:|----------:|--------:|-------------:|
 TopologicalSortCurrent |    1 |         1,108.2 ns |   1.00 |    0.3719 |       - |      1,176 B |
 TopologicalSortReduced |    1 |           843.5 ns |   0.76 |    0.2642 |       - |        832 B |
  TopologicalSortLookup |    1 |         1,084.2 ns |   0.98 |    0.3300 |       - |      1,040 B |
  
 TopologicalSortCurrent |    2 |         2,008.8 ns |   1.00 |    0.5684 |       - |      1,792 B |
 TopologicalSortReduced |    2 |         1,408.5 ns |   0.70 |    0.3452 |       - |      1,088 B |
  TopologicalSortLookup |    2 |         2,066.4 ns |   1.03 |    0.4196 |       - |      1,328 B |
  
 TopologicalSortCurrent |    4 |         4,549.4 ns |   1.00 |    1.0452 |       - |      3,312 B |
 TopologicalSortReduced |    4 |         2,760.1 ns |   0.61 |    0.4768 |       - |      1,512 B |
  TopologicalSortLookup |    4 |         3,845.4 ns |   0.85 |    0.6638 |       - |      2,096 B |
  
 TopologicalSortCurrent |    8 |        13,930.5 ns |   1.00 |    2.5177 |       - |      7,944 B |
 TopologicalSortReduced |    8 |         6,906.7 ns |   0.50 |    0.9384 |       - |      2,976 B |
  TopologicalSortLookup |    8 |         8,149.6 ns |   0.59 |    1.3275 |       - |      4,184 B |
  
 TopologicalSortCurrent |   16 |        59,967.4 ns |   1.00 |    6.8970 |       - |     21,760 B |
 TopologicalSortReduced |   16 |        25,544.5 ns |   0.43 |    1.8921 |       - |      6,000 B |
  TopologicalSortLookup |   16 |        19,502.9 ns |   0.33 |    2.3193 |       - |      7,336 B |
  
 TopologicalSortCurrent |   32 |       344,414.7 ns |   1.00 |   21.4844 |       - |     67,768 B |
 TopologicalSortReduced |   32 |       131,691.9 ns |   0.38 |    3.6621 |       - |     11,744 B |
  TopologicalSortLookup |   32 |        57,389.3 ns |   0.17 |    4.5776 |       - |     14,552 B |
  
 TopologicalSortCurrent |   64 |     2,337,703.0 ns |   1.00 |   70.3125 |       - |    233,456 B |
 TopologicalSortReduced |   64 |       839,191.2 ns |   0.36 |    6.8359 |       - |     23,064 B |
  TopologicalSortLookup |   64 |       213,013.0 ns |   0.09 |    9.0332 |       - |     29,184 B |
  
 TopologicalSortCurrent |  512 | 1,073,799,023.3 ns |   1.00 | 4062.5000 | 62.5000 | 12,874,392 B |
 TopologicalSortReduced |  512 |   330,733,852.1 ns |   0.31 |         - |       - |    181,016 B |
  TopologicalSortLookup |  512 |    20,610,365.5 ns |   0.02 |   62.5000 |       - |    241,360 B |
```

File: src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs
Comment (FransBouma) [inline src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: I think you meant 'CalculateRelationships', not 'CalcuateRelationships' ;) :P

Comment (benaadams) [reply src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs]: fixed

**Comment by benaadams** [comment]
Allocations reduced back; think am done 😄 

For 1024 dependencies its x 90 times faster and only uses 0.7% of the memory
```
                 Method | Deps |        Mean | Scaled |    Gen 0 |  Gen 1 | Gen 2 |    Allocated |
----------------------- |----- |------------:|-------:|---------:|-------:|------:|-------------:|
 TopologicalSortCurrent | 1024 | 8,047.69 ms |   1.00 | 16125.00 | 768.52 | 62.50 | 49,720.89 KB |
 TopologicalSortReduced | 1024 | 2,018.55 ms |   0.25 |    62.50 |      - |     - |    353.02 KB |
  TopologicalSortLookup | 1024 |    88.84 ms |   0.01 |   125.00 |      - |     - |    476.80 KB |
 TopologicalSortLookup2 | 1024 |    89.90 ms |   0.01 |   125.00 |      - |     - |    388.15 KB |
```

**Comment by sharwell** [comment]
@benaadams Thanks for all the work on this. If your Allocated numbers are accurate, then your final benchmark is still 1/30 the size of what Roslyn actually demands. 😨

**Comment by benaadams** [comment]
@sharwell Just worked out how to profile nuget; so can test real work results 😄 

`nuget restore Roslyn.sln -NoCache` allocates 38,754 MB 😱 

`TopologicalSortUtility` is 4,833 MB of that; taking 21,577ms (21.5 secs)

After this change (and latest branch with "Reduce boxing of LibraryType" https://github.com/NuGet/NuGet.Client/pull/1582)

`nuget restore Roslyn.sln -NoCache` allocates 30,064 MB

`TopologicalSortUtility` is 16 MB of that; taking 338ms

Allocs after/before


Time after/before

**Comment by sharwell** [comment]
And now you have a better understanding of the context of my bug reports 😆

**Comment by emgarten** [comment]
I'll run this branch through the CI to verify all functional tests are still passing with this before merging.

@benaadams `-NoCache` in your examples above will only impact restore if it is downloading packages, not if they are already in the global packages folder. To profile the downloads also you can run `NuGet.exe locals -clear all` first, or set the packages directory to an empty folder.

**Comment by emgarten** [comment]
This has a couple test failures, I'll debug to see what is happening.

**Comment by benaadams** [comment]
See the failure

**Comment by emgarten** [comment]
@benaadams I was able to track it down, `UpdateChildCounts` was modifying decreasing the parent count on parent packages instead of children. I think the naming caused that confusion.

Here is the fix: https://github.com/NuGet/NuGet.Client/commit/a778caadae0df3cf6c8f166c9b5c2e60b23649d8

And some unit tests for this utility method (nuget really should have had these before): https://github.com/NuGet/NuGet.Client/commit/e2997c1c537e8a133a90f230108cc8b3992e8bdf

**Comment by benaadams** [comment]
You're good! 😁

**Comment by benaadams** [comment]
My git-fu isn't up to check picking from a remote branch; are you ok using your branch?

**Comment by emgarten** [comment]
That works, I'll let the CI verify everything and then I'll merge it in.

**Comment by emgarten** [comment]
I've merged this, it will be in NuGet 4.4.0. Thanks for the fix @benaadams! 🚀

**Pull request merged**
